android/sdl_android/src/main/java/com/smartdevicelink/transport/utl/SdlDeviceListener.java

```diff
@@ -0,0 +1,270 @@
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+
+package com.smartdevicelink.transport.utl;
+
+import android.bluetooth.BluetoothDevice;
+import android.content.Context;
+import android.content.SharedPreferences;
+import android.os.Handler;
+import android.os.Looper;
+import android.os.Message;
+import android.support.annotation.NonNull;
+
+import com.smartdevicelink.transport.MultiplexBaseTransport;
+import com.smartdevicelink.transport.MultiplexBluetoothTransport;
+import com.smartdevicelink.transport.SdlRouterService;
+import com.smartdevicelink.util.DebugTool;
+import com.smartdevicelink.util.SdlAppInfo;
+
+import java.lang.ref.WeakReference;
+import java.util.List;
+
+
+public class SdlDeviceListener {
+
+    private static final String TAG = "SdlListener";
+    private static final int MIN_VERSION_REQUIRED = 13;
+    private static final String SDL_DEVICE_STATUS_SHARED_PREFS = "sdl.device.status";
+    private static final Object LOCK = new Object(), RUNNING_LOCK = new Object();
+
+    private final WeakReference<Context> contextWeakReference;
+    private final Callback callback;
+    private final BluetoothDevice connectedDevice;
+    private MultiplexBluetoothTransport bluetoothTransport;
+    private TransportHandler bluetoothHandler;
+    private Handler timeoutHandler;
+    private Runnable timeoutRunner;
+    private boolean isRunning = false;
+
+
+    public SdlDeviceListener(Context context, BluetoothDevice device, Callback callback) {
+        this.contextWeakReference = new WeakReference<>(context);
+        this.connectedDevice = device;
+        this.callback = callback;
+    }
+
+    /**
+     * This will start the SDL Device Listener with two paths. The first path will be a check
+     * against the supplied bluetooth device to see if it has already successfully connected as an
+     * SDL device. If it has, the supplied callback will be called immediately. If the device hasn't
+     * connected as an SDL device before, the SDL Device Listener will then open up an RFCOMM channel
+     * using the SDL UUID and await a potential connection. A timeout is used to ensure this only
+     * listens for a finite amount of time. If this is the first time the device has been seen, this
+     * will listen for 30 seconds, if it is not, this will listen for 15 seconds instead.
+     */
+    public void start() {
+        if (hasSDLConnected(contextWeakReference.get(), connectedDevice.getAddress())) {
```

Comment:
if I install the app while BT is already on, the `connectedDevice` comes as null so the app crashes with this exception
```
java.lang.NullPointerException: Attempt to invoke virtual method 'java.lang.String android.bluetooth.BluetoothDevice.getAddress()' on a null object reference
        at com.smartdevicelink.transport.utl.SdlDeviceListener.start(SdlDeviceListener.java:88)
        at com.smartdevicelink.transport.SdlBroadcastReceiver$2.onComplete(SdlBroadcastReceiver.java:298)
        at com.smartdevicelink.util.ServiceFinder.onFinished(ServiceFinder.java:124)
        at com.smartdevicelink.util.ServiceFinder.access$000(ServiceFinder.java:58)
        at com.smartdevicelink.util.ServiceFinder$1.run(ServiceFinder.java:84)
        at android.os.Handler.handleCallback(Handler.java:883)
        at android.os.Handler.dispatchMessage(Handler.java:100)
        at android.os.Looper.loop(Looper.java:214)
        at android.app.ActivityThread.main(ActivityThread.java:7356)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930)
```

Reply:
Bluetooth had to be on and connected for me to see this error, but added a null check before this now so it should throw an error. It's not an easy task to get the connected bluetooth device, so I'm not sure if it is really worth the effort at the moment.